### PR TITLE
Add `--toolchain` option

### PR DESCRIPTION
**Pull request by tkrajacic**
Adds the ability to specify a custom Swift toolchain via a `--toolchain` option on the `build`, `update`, and `archive` commands.

E.g.: `carthage build --toolchain org.swift.3020160509a`

**Todo**
- [x] Make toolchain parameter default to `nil` where desired
- [x] tests?

**Resolves issue** #1039

### CODE REVIEW

**Comment by tkrajacic** [comment]
Since I haven't updated the tests, and the function arguments require the `toolchain` parameter at the moment, ci fails of course.
I'll update that after initial feedback if the general approach is fine.

File: Source/CarthageKit/Xcode.swift [hunk 1]
```diff
@@ -171,14 +171,14 @@ public func schemesInProject(project: ProjectLocator) -> SignalProducer<String,
...
-public func buildableSchemesInDirectory(directoryURL: NSURL, withConfiguration configuration: String, forPlatforms platforms: Set<Platform> = []) -> SignalProducer<(ProjectLocator, [String]), CarthageError> {
+public func buildableSchemesInDirectory(directoryURL: NSURL, withConfiguration configuration: String, forPlatforms platforms: Set<Platform> = [], toolchain: String?) -> SignalProducer<(ProjectLocator, [String]), CarthageError> {
 	precondition(directoryURL.fileURL)
 
 	return locateProjectsInDirectory(directoryURL)
 		.flatMap(.Concat) { project -> SignalProducer<(ProjectLocator, [String]), CarthageError> in
 			return schemesInProject(project)
 				.flatMap(.Merge) { scheme -> SignalProducer<String, CarthageError> in
-					let buildArguments = BuildArguments(project: project, scheme: scheme, configuration: configuration)
+					let buildArguments = BuildArguments(project: project, scheme: scheme, configuration: configuration, toolchain: toolchain)
```
Comment (mdiep) [inline Source/CarthageKit/Xcode.swift]: Is this actually needed? It doesn't seem like the toolchain would affect whether we'd build the scheme.

File: Source/carthage/Archive.swift [hunk 2]
```diff
@@ -14,19 +14,21 @@ import ReactiveCocoa
...
-			} } }
+		static func create(outputPath: String?) -> String? -> String -> ColorOptions -> [String] -> Options {
+			return { toolchain in { directoryPath in { colorOptions in { frameworkNames in
+				return self.init(toolchain: toolchain, outputPath: outputPath, directoryPath: directoryPath, colorOptions: colorOptions, frameworkNames: frameworkNames)
+			} } } }
 		}
 
 		public static func evaluate(m: CommandMode) -> Result<Options, CommandantError<CarthageError>> {
 			return create
+				<*> m <| Option<String?>(key: "toolchain", defaultValue: nil, usage: "the toolchain to build with")
```
Comment (mdiep) [inline Source/carthage/Archive.swift]: `carthage archive` doesn't build anything, so it shouldn't need the `toolchain` option.

**Comment by mdiep** [comment]
This is a great first step! ☺️

**Comment by tkrajacic** [comment]
I removed the unnecessary `toolchain` parameters. They were remnants from when I was hunting down compile errors ;)
I also added a default `nil` to the `BuildArgument` constructor and removed all explicit `nil` passing.

**Comment by tkrajacic** [comment]
I didn't add the default `nil` to `buildDependencyProject(...)` and `buildInDirectory(...)` because _CarthageKit_ always passes in the toolchain from a higher level command.

When adding an expectation to the _BuildArgumentsSpec_ though I saw that the test don't do that and call those methods directly.

Does it make sense to add the default `nil` value to the `toolchain` parameter as well on `buildDependencyProject(...)` and `buildInDirectory(...)`? In my mind it doesn't make sense to leave out the toolchain anywhere downstream because once it has been supplied as an argument it must be used everywhere (all build products need to be compiled with the same toolchain).

**Comment by mdiep** [comment]
Let's add `nil` defaults to all the `public` functions that take a toolchain. That way we don't break compatibility for CarthageKit.

Also, I missed this yesterday, but `carthage bootstrap` needs a `--toolchain` argument as well.

This is shaping up great!

**Comment by tkrajacic** [comment]
Doesn't `bootstrap` already work with all of `update`s commands?
It reuses the `UpdateCommand.Options`, and running `bootstrap --toolchain xxx` did work when I tried it.

**Comment by mdiep** [comment]
> Doesn't bootstrap already work with all of updates commands?
> It reuses the UpdateCommand.Options, and running bootstrap --toolchain xxx did work when I tried it.

Oh, duh!

**Comment by mdiep** [comment]
This is great! Thanks so much for the pull request! ✨

**Comment by NachoSoto** [comment]
What's the expected value for this parameter? I'm trying "swift-DEVELOPMENT-SNAPSHOT-2016-09-06-a", as well as the full path, but Carthage is ignoring it and using the default toolchain.

**Comment by tkrajacic** [comment]
For the tests I had `org.swift.3020160509a` but I don't immediately recall how I found this. (I am already in bed over here so I can't check right now)

I'll see what I can find in the morning.

**Comment by ikesyo** [comment]
> What's the expected value for this parameter?

The `CFBundleIdentifier` value of `Info.plist` located at `/Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2016-MM-dd-a.xctoolchain/`.

**Pull request merged**